src/components/ImagePreviewer.js

```diff
@@ -0,0 +1,176 @@
+import { stringify } from "querystring";
+import { decode } from "base58";
+
+const noop = () => {};
+
+export const of = src => Promise.resolve({ src });
+
+export const makeSrcResolver = ({ src }) =>
+  resolvers.reduceRight((acc, resolver) => {
+    if (!acc && resolver.test(src)) {
+      return resolver.request(src);
+    }
+    return acc;
+  }, false);
+
+export const makeImageDOMResolver = ({ src }) =>
+  new Promise((resolve, reject) => {
+    const img = new Image();
+    img.onload = () =>
+      resolve({
+        src,
+        height: img.height
+      });
+    img.onerror = reject;
+    img.src = src;
+  });
+
+export class ImagePreviewer extends React.PureComponent {
+  state = {
+    pending: undefined,
+    value: undefined
+  };
+
+  componentDidMount() {
+    this.handleStart();
+  }
+
+  componentDidUpdate(prevProps) {
+    if (this.props.request !== prevProps.request) {
+      this.handleStart();
+    }
+  }
+
+  handleStart(props) {
+    this.setState((state, { request }) => {
+      request.then(this.handleResolve, noop);
+      return {
+        pending: request,
+        value: undefined
+      };
+    });
+  }
+
+  handleResolve = value => {
+    this.setState(({ pending }, { request }) => {
+      if (pending !== request) {
+        return;
+      }
+      return { value };
+    });
+  };
+
+  render() {
+    return this.props.render({
+      ...this.props,
+      request: undefined,
+      ...this.state.value
```

Comment:
I feel like it should explicit say which prop to pass rather than this `...` hack and remove prop using `request: undefined`.

Reply:
The `...` operator is used widely and it's just the equivalent of `Object.assign`. Again, it keeps flexibility of defining react components/render functions here